File: go-sn.c
#define _GNU_SOURCE //enable the strcasestr function from string.h

#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdint.h>
#include <string.h>
#include <errno.h>
#include <sys/stat.h>
#include <sys/types.h>

#define BLOCK_SIZE 512
#define SN_LEN 19
#define SERIAL_PARTITION "/dev/disk/by-partlabel/serial"
#define LEGACY_DEVICE "/dev/mmcblk0"

#ifndef VERSION
#error "VERSION must be passed in by command: -DVERSION=<version>"
#endif

#define stringify_(x) #x
#define stringify(x) stringify_(x)

const char* usage =
"GOcontroll serial number utility v" stringify(VERSION) "\n"
"Usage:\n"
"go-sn [command] \'serial-number\'\n"
"\n"
"Available commands:\n"
"[r]ead		Read the serial number from memory\n"
"[w]rite		Write the 'serial-number' to memory\n"
"\n"
"'serial-number' must be a total of 19 characters long, and be segmented into 4 parts of 4 seperated by '-'\n"
"\n"
"Storage: prefers GPT partition labelled 'serial' (modern firmware),\n"
"falls back to last sector of /dev/mmcblk0 (legacy units pre-serial-partition).\n"
"\n"
"Examples:\n"
"go-sn read\n"
"go-sn write 1234-5678-9012-3456\n";

int validate_serial(char* sn) {
	if (strlen(sn) != SN_LEN) {
		fprintf(stderr, "The serial number has to be 19 characters long.\n");
		return -1;
	}
	for (int i = 0; i < SN_LEN; i++) { // sn is already verified to be SN_LEN long
		if ((sn[i] == '-') && (((i+1) % 5) != 0)) { //check if every - is at the proper position
			fprintf(stderr, "Each segment of the of the serial number should contain 4 characters\n");
			return -1;
		}
		if ((sn[i] != '-') && (((i+1) % 5) == 0)) { //check if every - position contains a -
			fprintf(stderr, "Each segment of the of the serial number should contain 4 characters\n");
			return -1;
		}
	}
	return 0;
}

// Open /dev/mmcblk0 and seek to the last sector. Used as legacy fallback when
// the dedicated GPT serial partition is not present (pre-migration firmware).
static int open_legacy(int flags) {
	int disk = open(LEGACY_DEVICE, flags);
	if (disk < 0) {
		fprintf(stderr, "Unable to open block device:\n%s\n", strerror(errno));
		return -1;
	}
	if (lseek(disk, -BLOCK_SIZE, SEEK_END) < 0) {
		close(disk);
		fprintf(stderr, "Could not seek to legacy serial number address:\n%s\n", strerror(errno));
		return -1;
	}
	return disk;
}

int read_serial() {
	char buffer[SN_LEN+1]; //buffer with null terminator

	// Try the dedicated GPT serial partition first.
	int disk = open(SERIAL_PARTITION, O_RDONLY);
	if (disk >= 0) {
		ssize_t bytes = read(disk, buffer, SN_LEN);
		close(disk);
		if (bytes == SN_LEN) {
			buffer[SN_LEN] = 0;
			if (validate_serial(buffer) == 0) {
				printf("%s\n", buffer);
				return 0;
			}
			// Partition is present but contains no valid SN (freshly flashed
			// unit, or migration-from-legacy). Fall through to legacy probe.
		}
	} else if (errno != ENOENT) {
		fprintf(stderr, "Unable to open serial partition:\n%s\n", strerror(errno));
		return -1;
	}

	// Legacy fallback.
	disk = open_legacy(O_RDONLY);
	if (disk < 0) {
		return -1;
	}
	ssize_t bytes = read(disk, buffer, SN_LEN);
	close(disk);
	if (bytes < 0) {
		fprintf(stderr, "Could not read the serial number:\n%s\n", strerror(errno));
		return -1;
	}
	if (bytes != SN_LEN) {
		fprintf(stderr, "Could not read full serial number, only %zd bytes read.\n", bytes);
		return -1;
	}
	buffer[SN_LEN] = 0;
	if (validate_serial(buffer) != 0) {
		return -1;
	}
	printf("%s\n", buffer); //print the sn to stdout
	return 0;
}

int write_serial(char* sn) {
	if (validate_serial(sn)) {
		return -1;
	}
	// Pad to a full sector with SN at offset 0, zeros after — block-device
	// friendly (avoids partial-sector read-modify-write through buffer cache).
	char buffer[BLOCK_SIZE] = {0};
	strncpy(buffer, sn, SN_LEN);

	// Prefer the dedicated GPT serial partition.
	int disk = open(SERIAL_PARTITION, O_RDWR);
	if (disk >= 0) {
		ssize_t bytes = write(disk, buffer, BLOCK_SIZE);
		close(disk);
		if (bytes != BLOCK_SIZE) {
			fprintf(stderr, "Could not write full block to serial partition.\n");
			return -1;
		}
		return 0;
	}
	if (errno != ENOENT) {
		fprintf(stderr, "Unable to open serial partition:\n%s\n", strerror(errno));
		return -1;
	}

	// Legacy fallback. Note: on modern firmware (with serial partition)
	// this region is the GPT backup header — we never get here in that
	// case because the partition open above succeeded.
	disk = open_legacy(O_RDWR);
	if (disk < 0) {
		return -1;
	}
	ssize_t bytes = write(disk, buffer, BLOCK_SIZE);
	close(disk);
	if (bytes != BLOCK_SIZE) {
		fprintf(stderr, "Could not write full block to legacy location.\n");
		return -1;
	}
	return 0;
}

int main (int argc, char *argv[]) {
	if (argc < 2) {
		fprintf(stderr,"%s",usage);
		return 0;
	}
	if (strcasestr("read", argv[1]) != NULL) {
		return read_serial();
	} else if (strcasestr("write", argv[1]) != NULL) {
		if (argc != 3) {
			fprintf(stderr,"%s", usage);
			return -1;
		}
		return write_serial(argv[2]);
	}
}

File: go-sn.zig
const std = @import("std");
const config = @import("config");

const BLOCK_SIZE = 512;
const SN_LEN = 19;
const SERIAL_PARTITION = "/dev/disk/by-partlabel/serial";
const LEGACY_DEVICE = "/dev/mmcblk0";

const usage =
    \\GOcontroll serial numer utility v{s}
    \\Usage:
    \\go-sn [command] 'serial-number'
    \\
    \\Available commands:
    \\[r]ead		Read the serial number from memory
    \\[w]rite		Write the 'serial-number' to memory
    \\
    \\'serial-number' must be a total of 19 characters long, and be segmented into 4 parts of 4 seperated by '-'
    \\
    \\Storage: prefers GPT partition labelled 'serial' (modern firmware),
    \\falls back to last sector of /dev/mmcblk0 (legacy units pre-serial-partition).
    \\
    \\Examples:
    \\go-sn read
    \\go-sn write 1234-5678-9012-3456
    \\
;

pub fn main() !void {
    var args = std.process.args();
    if (!args.skip()) {
        std.debug.print(usage, .{config.version});
        return;
    }
    const command = args.next() orelse {
        std.debug.print(usage, .{config.version});
        return;
    };

    if (std.ascii.startsWithIgnoreCase("read", command)) {
        try read_sn();
    } else if (std.ascii.startsWithIgnoreCase("write", command)) {
        const sn = args.next() orelse {
            std.log.err(usage, .{config.version});
            return error.InvalidArgument;
        };
        try write_sn(sn);
    } else {
        std.log.err(usage, .{config.version});
        return error.InvalidArgument;
    }
}

fn read_sn() !void {
    var sn: [SN_LEN]u8 = undefined;

    // Try the dedicated GPT serial partition first.
    if (std.fs.openFileAbsolute(SERIAL_PARTITION, .{ .mode = .read_only })) |f| {
        defer f.close();
        const bytes = try f.read(&sn);
        if (bytes == SN_LEN) {
            if (validate_sn(&sn)) |_| {
                return print_sn(&sn);
            } else |_| {
                // Partition is present but contains no valid SN (freshly flashed
                // unit, or migration-from-legacy). Fall through to legacy probe.
            }
        }
    } else |err| {
        if (err != error.FileNotFound) return err;
    }

    // Legacy fallback: last sector of /dev/mmcblk0 (pre-serial-partition firmware).
    const f = try std.fs.openFileAbsolute(LEGACY_DEVICE, .{ .mode = .read_only });
    defer f.close();
    try f.seekFromEnd(-BLOCK_SIZE);
    const bytes = try f.read(&sn);
    if (bytes != SN_LEN) {
        std.log.err("Could not read full serial number, only {} bytes read.\n", .{bytes});
        return error.IOError;
    }
    try validate_sn(&sn);
    try print_sn(&sn);
}

fn write_sn(sn: []const u8) !void {
    try validate_sn(sn);

    // Pad to a full sector with SN at offset 0, zeros after — block-device
    // friendly (avoids partial-sector read-modify-write through the buffer cache).
    var buffer = [_]u8{0} ** BLOCK_SIZE;
    @memcpy(buffer[0..SN_LEN], sn);

    // Prefer the dedicated GPT serial partition.
    if (std.fs.openFileAbsolute(SERIAL_PARTITION, .{ .mode = .read_write })) |f| {
        defer f.close();
        try f.writeAll(&buffer);
        return;
    } else |err| {
        if (err != error.FileNotFound) return err;
    }

    // Legacy fallback: last sector of /dev/mmcblk0. Note: on modern firmware
    // (with serial partition) this region is the GPT backup header — we never
    // get here in that case because the partition open above succeeded.
    const f = try std.fs.openFileAbsolute(LEGACY_DEVICE, .{ .mode = .read_write });
    defer f.close();
    try f.seekFromEnd(-BLOCK_SIZE);
    try f.writeAll(&buffer);
}

fn print_sn(sn: []const u8) !void {
    const stdout_file = std.io.getStdOut().writer();
    var bw = std.io.bufferedWriter(stdout_file);
    const stdout = bw.writer();
    try stdout.print("{s}\n", .{sn});
    try bw.flush();
}

fn validate_sn(sn: []const u8) !void {
    if (sn.len != SN_LEN) {
        std.log.err("The serial number has to be 19 characters long.\n", .{});
        return error.InvalidArgument;
    }
    var segments = std.mem.splitSequence(u8, sn, "-");
    while (segments.next()) |segment| {
        if (segment.len != 4) {
            std.log.err("Each segment of the of the serial number should contain 4 characters\n", .{});
            return error.InvalidArgument;
        }
    }
}
